osu.Game/Database/RealmAccess.cs

```diff
@@ -605,16 +586,33 @@ public IDisposable BlockAllOperations()
                             throw new InvalidOperationException(@$"{nameof(BlockAllOperations)} must be called from the update thread.");
 
                         syncContext = SynchronizationContext.Current;
-                    }
 
-                    unregisterAllSubscriptions();
+                        // Before disposing the update context, clean up all subscriptions.
+                        // Note that in the case of realm notification subscriptions, this is not really required (they will be cleaned up by disposal).
+                        // In the case of custom subscriptions, we want them to fire before the update realm is disposed in case they do any follow-up work.
+                        foreach (var action in customSubscriptionsResetMap)
+                        {
+                            action.Value?.Dispose();
+                            customSubscriptionsResetMap[action.Key] = null;
+                        }
+                    }
 
                     Logger.Log(@"Blocking realm operations.", LoggingTarget.Database);
 
                     updateRealm?.Dispose();
                     updateRealm = null;
                 }
 
+                // In order to ensure events arrive in the correct order, these *must* be fired post disposal of the update realm,
+                // and must be posted to the synchronization context.
+                // This is because realm may fire event callbacks between the `unregisterAllSubscriptions` and `updateRealm.Dispose`
+                // calls above.
+                syncContext?.Send(_ =>
+                {
+                    foreach (var action in notificationsResetMap.Values)
+                        action();
```

Comment:
This is the only usage of `notificationResetMap` which doesn't take `realmLock`. Same as last time, it probably doesn't strictly matter as all of the surrounding code is pretty much only allowed to run on the update thread, but given that this sort of thing was pointed out in review last time [no less](https://github.com/ppy/osu/pull/16547#discussion_r790278090) [than twice](https://github.com/ppy/osu/pull/16547#discussion_r790505471) then maybe better to just lock for the sake of it?